Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -64,45 +63,212 @@ extension PluginTarget {
+                            }
+                        }
+                        catch {
+                            print("couldn't send reply to plugin \(error)")
+                        }
+                    }
+
+                case .symbolGraphRequest(let targetName, let options):
+                    // The plugin requested symbol graph information for a target. We ask the delegate and then send a response.
+                    delegate.pluginRequestedSymbolGraph(forTarget: .init(targetName), options: .init(options)) {
+                        do {
+                            switch $0 {
+                            case .success(let result):
+                                responder(try HostToPluginMessage.symbolGraphResponse(result: .init(result)).toData())
+                            case .failure(let error):
+                                responder(try HostToPluginMessage.errorResponse(error: String(describing: error)).toData())
+                            }
+                        }
+                        catch {
+                            print("couldn't send reply to plugin \(error)")
+                        }
+                    }
+                }
+            }
+        }
+        let runnerDelegate = ScriptRunnerDelegate(callbackQueue: callbackQueue, delegate: delegate)
+        
         // Call the plugin script runner to actually invoke the plugin.
         scriptRunner.runPluginScript(
             sources: sources,
-            input: inputStruct,
+            initialMessage: initialMessage,
             toolsVersion: self.apiVersion,
             workingDirectory: workingDirectory,
             writableDirectories: writableDirectories,
             readOnlyDirectories: readOnlyDirectories,
             fileSystem: fileSystem,
             observabilityScope: observabilityScope,
             callbackQueue: callbackQueue,
-            delegate: delegate,
-            completion: completion)
+            delegate: runnerDelegate) { result in
+                dispatchPrecondition(condition: .onQueue(callbackQueue))
+                switch result {
+                case .success(let exitCode):
+                    // Return a result based on the exit code. If the plugin
+                    // exits with an error but hasn't already emitted an error,
+                    // we do so for it.
+                    let success = (exitCode == 0)
+                    if !success && runnerDelegate.errorCount == 0 {
+                        delegate.pluginEmittedDiagnostic(
+                            .error("Plugin ended with exit code \(exitCode)")
+                        )
+                    }
+                    completion(.success(exitCode == 0))
+                case .failure(let error):
+                    completion(.failure(error))
+                }
+        }
+    }
+}
+
+fileprivate extension HostToPluginMessage {
+    func toData() throws -> Data {
+        let encoder = JSONEncoder()
```

Comment:
makeWithDefaults?

Reply:
Ah right, I keep forgetting about that one.  :)  Thanks for pointing it out.